### DISCUSSION TITLE
Issues in Testing FastAPI with existing DB: Instance xxxx is not persistent within this Session

### DISCUSSION THREAD

**Discussion by cesarqdt** [fastapi/fastapi#12779] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# conftest.py
engine = create_engine(connection_url)

@pytest.fixture(name = "session")
def session_fixture():
    with Session(engine) as session:
        yield session

@pytest.fixture(name = "client")
def client_fixture(session: Session):
    def get_db_override():
        return session

    app = create_app()
    app.dependency_overrides[get_db] = get_db_override

    client = TestClient(app)
    yield client
    app.dependency_overrides.clear()

# tests
from fastapi import status
from fastapi.testclient import TestClient


def test_link_app_to_user(client: TestClient, superuser_token, normal_user_id, test_app_id):
    response = client.post(
        f"/apps/{test_app_id}/users/{normal_user_id}",
        headers = {"Authorization": f"Bearer {superuser_token}"},
    )
    if response.status_code == status.HTTP_409_CONFLICT:
        data = response.json()
        assert "detail" in data
        assert data["detail"] == "The user already has permission to use this app."
        return

    assert response.status_code == status.HTTP_200_OK
    data = response.json()
    assert "message" in data
    assert data["message"] == "App linked to user successfully"
    client.close()

def test_unlink_app_to_user(client: TestClient, superuser_token, normal_user_id, test_app_id):
    response = client.delete(
        f"/apps/{test_app_id}/users/{normal_user_id}",
        headers = {"Authorization": f"Bearer {superuser_token}"},
    )
    if response.status_code == status.HTTP_409_CONFLICT:
        data = response.json()
        assert "detail" in data
        assert data["detail"] == "The user does not have permission to use this app."
        return

    assert response.status_code == status.HTTP_200_OK
    data = response.json()
    assert "message" in data
    assert data["message"] == "App unlinked from user successfully"
```

### Description

I am testing a FastAPI application with pytest, using a live connection to an SQLModel database. I have setup issues where consecutive database operations (like POST followed by DELETE) in the same test suite fail, with errors indicating session inconsistencies. Specifically, I encounter errors like:
```
FAILED authentication/tests/api/test_links.py::test_unlink_app_to_user - sqlalchemy.exc.InvalidRequestError: Instance '<UserAppRole at 0x10a12d680>' is not persistent within this Session
```
This test fails in the second test of the script which is the `test_unlinkl_app_to_user` function. I already followed the instructions here: https://sqlmodel.tiangolo.com/tutorial/fastapi/tests/#memory-database
I don't want to mock the database, because I am testing with existing objects. 

At this point I am not sure what I'm doing wrong... My assumption is that the client and therefore the session is created per test, so no matter what I do in one test, I should be able to do the operation in the second, and third, etc.

### Operating System

macOS

### FastAPI Version

0.111.1

### Pydantic Version

2.9.2

### Python Version

Python 3.11.4

**Comment by luzzodev** [comment]
In pytest fixture have a scope with default to function.
If the intenction is to to share the session and the objects in it between multiple tests, you need the adjust the scope of your fixture.
More informations on [Here](https://docs.pytest.org/en/6.2.x/fixture.html) under chapter: 

> Scope: sharing fixtures across classes, modules, packages or session

**Comment by cesarqdt** [comment]
@luzzodev I've adjusted the scope to be `session` as below:
```
@pytest.fixture(name = "session", scope = "session")
def session_fixture():
    with Session(engine) as session:
        yield session

@pytest.fixture(name = "client", scope = "session")
def client_fixture(session: Session):
    def get_db_override():
        return session

    app = create_app()
    app.dependency_overrides[get_db] = get_db_override

    client = TestClient(app)
    yield client
    app.dependency_overrides.clear()
```
But this is still giving me the same error `FAILED authentication/tests/api/test_links.py::test_unlink_app_to_user - sqlalchemy.exc.InvalidRequestError: Instance '<UserAppRole at 0x10ddc0e10>' is not persistent within this Session`.

I checked that the client is the same for both tests, which it is `client = <starlette.testclient.TestClient object at 0x10dd60e90>` from `test_link_app_to_user` and `client = <starlette.testclient.TestClient object at 0x10dd60e90>`.

**Reply by luzzodev** [reply]
can you provide the full example, including the dependency functions and the routes you are trying to test. So i can replicate the example and help you

**Reply by cesarqdt** [reply, marked as answer]
Thanks for your help @luzzodev I realized something very silly in my code that was the root cause of this issue. In fact, the session as defined in the fixture along with the scope should have fixed my problem, but it did not because of the CRUD operations I was doing in the endpoint itself.

I was doing a delete operation, which in the crud function had a commit() and after that operation, I was doing another delete operation and a refresh() afterwards, which was obviously a wrong pattern in the app.

The way I fixed it was by keep the session as it was, but then removing the commits from the crud functions and moving them out to the endpoint (caller), something like this:
```python
# crud.py
def delete_first_row(session, ...):
    session.delete(...)

def delete_second_row(session, ...):
    session.delete(...)

# apis
def endpoint(session, ....):
    crud.delete_first_row(session, ....)
    crud.delete_second(session, ...)
    session.commit()
```

After this change, I was able to run the operation properly (and the pytests as well).
